File: clients/ios/FamilyAI/Core/Networking/GatewayClient.swift
import Foundation

#if canImport(FoundationNetworking)
  import FoundationNetworking
#endif

public struct GatewayProfile: Codable, Equatable, Sendable {
  public let baseURL: URL
  public init(baseURL: URL) { self.baseURL = baseURL }
}
public struct DeviceAuthorization: Codable, Equatable, Sendable {
  public let deviceRef: String
  public let deviceCredential: String
  public init(deviceRef: String, deviceCredential: String) {
    self.deviceRef = deviceRef
    self.deviceCredential = deviceCredential
  }
}
public protocol GatewayClientProtocol: Sendable {
  func preview(baseURL: URL, request: PairingPreviewRequest) async throws -> PairingPreviewResponse
  func claim(baseURL: URL, request: PairingClaimRequest) async throws -> PairingClaimResponse
  func fetchPortalContext(baseURL: URL, session: EntrySessionCredential) async throws
    -> PersonalPortalContext
  func renew(baseURL: URL, authorization: DeviceAuthorization) async throws -> SessionRenewResponse
  func logout(baseURL: URL, session: EntrySessionCredential) async throws -> MobileOperationResponse
  func unbind(baseURL: URL, authorization: DeviceAuthorization) async throws
    -> MobileOperationResponse
}
public final class URLSessionGatewayClient: GatewayClientProtocol, @unchecked Sendable {
  private enum Authentication {
    case none
    case entry(EntrySessionCredential)
    case device(DeviceAuthorization)
  }
  private let session: URLSession
  private let decoder = MobileEntryCoding.decoder()
  private let encoder = MobileEntryCoding.encoder()
  public init(session: URLSession? = nil) {
    if let session {
      self.session = session
    } else {
      let configuration = URLSessionConfiguration.ephemeral
      configuration.timeoutIntervalForRequest = 15
      configuration.timeoutIntervalForResource = 15
      self.session = URLSession(configuration: configuration)
    }
  }
  public func preview(baseURL: URL, request: PairingPreviewRequest) async throws
    -> PairingPreviewResponse
  {
    try await send(
      baseURL: baseURL, path: "/api/v1/mobile/pairing/preview", method: "POST", body: request,
      authentication: .none)
  }
  public func claim(baseURL: URL, request: PairingClaimRequest) async throws -> PairingClaimResponse
  {
    try await send(
      baseURL: baseURL, path: "/api/v1/mobile/pairing/claim", method: "POST", body: request,
      authentication: .none)
  }
  public func fetchPortalContext(baseURL: URL, session: EntrySessionCredential) async throws
    -> PersonalPortalContext
  {
    try await send(
      baseURL: baseURL, path: "/api/v1/portal/context", method: "GET", body: Optional<String>.none,
      authentication: .entry(session))
  }
  public func renew(baseURL: URL, authorization: DeviceAuthorization) async throws
    -> SessionRenewResponse
  {
    try await send(
      baseURL: baseURL, path: "/api/v1/mobile/session/renew", method: "POST",
      body: Optional<String>.none, authentication: .device(authorization))
  }
  public func logout(baseURL: URL, session: EntrySessionCredential) async throws
    -> MobileOperationResponse
  {
    try await send(
      baseURL: baseURL, path: "/api/v1/mobile/session/logout", method: "POST",
      body: Optional<String>.none, authentication: .entry(session))
  }
  public func unbind(baseURL: URL, authorization: DeviceAuthorization) async throws
    -> MobileOperationResponse
  {
    try await send(
      baseURL: baseURL, path: "/api/v1/mobile/device", method: "DELETE",
      body: Optional<String>.none, authentication: .device(authorization))
  }
  private func send<Response: Decodable, Body: Encodable>(
    baseURL: URL, path: String, method: String, body: Body?, authentication: Authentication
  ) async throws -> Response {
    guard let secureBaseURL = try? GatewayURLValidator.validate(baseURL.absoluteString),
      let url = URL(string: path, relativeTo: secureBaseURL)?.absoluteURL
    else { throw GatewayClientError.insecureGateway }
    var request = URLRequest(url: url, timeoutInterval: 15)
    request.httpMethod = method
    request.setValue("application/json", forHTTPHeaderField: "Accept")
    request.setValue("ios-\(UUID().uuidString.lowercased())", forHTTPHeaderField: "X-Request-ID")
    if let body {
      request.httpBody = try encoder.encode(body)
      request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    }
    switch authentication {
    case .none: break
    case .entry(let credential):
      request.setValue("Bearer \(credential.token)", forHTTPHeaderField: "Authorization")
      request.setValue(credential.entrySessionRef, forHTTPHeaderField: "X-Entry-Session-Ref")
    case .device(let authorization):
      request.setValue(
        "Device \(authorization.deviceCredential)", forHTTPHeaderField: "Authorization")
      request.setValue(authorization.deviceRef, forHTTPHeaderField: "X-Device-Ref")
    }
    do {
      let (data, response) = try await session.data(for: request)
      guard let http = response as? HTTPURLResponse else {
        throw GatewayClientError.invalidResponse
      }
      guard (200..<300).contains(http.statusCode) else {
        if let envelope = try? decoder.decode(MobileGatewayErrorEnvelope.self, from: data) {
          throw GatewayClientError.server(envelope.error.code)
        }
        throw GatewayClientError.invalidResponse
      }
      do { return try decoder.decode(Response.self, from: data) } catch {
        throw GatewayClientError.invalidResponse
      }
    } catch let error as GatewayClientError { throw error } catch let error as URLError {
      if error.code == .timedOut { throw GatewayClientError.timeout }
      if [
        .cannotConnectToHost, .cannotFindHost, .dnsLookupFailed, .networkConnectionLost,
        .notConnectedToInternet,
      ].contains(error.code) {
        throw GatewayClientError.unreachable
      }
      throw GatewayClientError.invalidResponse
    } catch { throw GatewayClientError.invalidResponse }
  }
}

File: clients/harmonyos/core/src/gatewayClient.ts
import { MOBILE_ENDPOINTS } from './endpoints.ts';
import { validateGatewayBaseUrl } from './gatewayUrl.ts';
import {
  buildDeviceRequest,
  buildEntryRequest,
  type DeviceAuthorization,
  type GatewayRequest
} from './requests.ts';
import type {
  EntrySessionCredential,
  MobileGatewayErrorCode,
  MobileOperationResponse,
  PersonalPortalContext,
  SessionRenewResponse
} from './types.ts';
import {
  parseMobileGatewayError,
  parseMobileOperationResponse,
  parsePersonalPortalContext,
  parseSessionRenewResponse
} from './validation.ts';

export interface GatewayProfile {
  baseURL: string;
}

export interface GatewayTransportResponse {
  status: number;
  body: unknown;
}

export interface GatewayTransport {
  send(input: {
    baseURL: string;
    request: GatewayRequest;
  }): Promise<GatewayTransportResponse>;
}

export type GatewayTransportErrorKind = 'timeout' | 'unreachable';

export class GatewayTransportError extends Error {
  readonly kind: GatewayTransportErrorKind;

  constructor(kind: GatewayTransportErrorKind) {
    super(kind);
    this.name = 'GatewayTransportError';
    this.kind = kind;
  }
}

export type GatewayClientErrorKind =
  | 'timeout'
  | 'unreachable'
  | 'invalid_response'
  | 'insecure_gateway'
  | 'server';

export class GatewayClientError extends Error {
  readonly kind: GatewayClientErrorKind;
  readonly serverCode: MobileGatewayErrorCode | undefined;

  constructor(
    kind: GatewayClientErrorKind,
    serverCode?: MobileGatewayErrorCode
  ) {
    super(kind === 'server' && serverCode ? `${kind}:${serverCode}` : kind);
    this.name = 'GatewayClientError';
    this.kind = kind;
    this.serverCode = serverCode;
  }
}

type ResponseParser<T> = (value: unknown) => T;

export class MobileGatewayClient {
  private readonly transport: GatewayTransport;

  constructor(transport: GatewayTransport) {
    this.transport = transport;
  }

  fetchPortalContext(
    profile: GatewayProfile,
    session: EntrySessionCredential
  ): Promise<PersonalPortalContext> {
    const endpoint = MOBILE_ENDPOINTS.portalContext;
    return this.send(
      profile,
      buildEntryRequest(endpoint.path, endpoint.method, {
        entrySessionRef: session.entrySessionRef,
        token: session.token
      }),
      parsePersonalPortalContext
    );
  }

  renew(
    profile: GatewayProfile,
    authorization: DeviceAuthorization
  ): Promise<SessionRenewResponse> {
    const endpoint = MOBILE_ENDPOINTS.sessionRenew;
    return this.send(
      profile,
      buildDeviceRequest(endpoint.path, endpoint.method, authorization),
      parseSessionRenewResponse
    );
  }

  logout(
    profile: GatewayProfile,
    authorization: DeviceAuthorization
  ): Promise<MobileOperationResponse> {
    const endpoint = MOBILE_ENDPOINTS.sessionLogout;
    return this.send(
      profile,
      buildDeviceRequest(endpoint.path, endpoint.method, authorization),
      parseMobileOperationResponse
    );
  }

  unbind(
    profile: GatewayProfile,
    authorization: DeviceAuthorization
  ): Promise<MobileOperationResponse> {
    const endpoint = MOBILE_ENDPOINTS.deviceUnbind;
    return this.send(
      profile,
      buildDeviceRequest(endpoint.path, endpoint.method, authorization),
      parseMobileOperationResponse
    );
  }

  private async send<T>(
    profile: GatewayProfile,
    request: GatewayRequest,
    parse: ResponseParser<T>
  ): Promise<T> {
    let baseURL: string;
    try {
      baseURL = validateGatewayBaseUrl(profile.baseURL);
    } catch {
      throw new GatewayClientError('insecure_gateway');
    }

    const requestWithAccept: GatewayRequest = {
      ...request,
      headers: {
        accept: 'application/json',
        ...request.headers
      }
    };

    let response: GatewayTransportResponse;
    try {
      response = await this.transport.send({
        baseURL,
        request: requestWithAccept
      });
    } catch (error) {
      if (error instanceof GatewayTransportError) {
        throw new GatewayClientError(error.kind);
      }
      if (error instanceof GatewayClientError) throw error;
      throw new GatewayClientError('invalid_response');
    }

    if (response.status >= 200 && response.status < 300) {
      try {
        return parse(response.body);
      } catch {
        throw new GatewayClientError('invalid_response');
      }
    }

    try {
      const envelope = parseMobileGatewayError(response.body);
      throw new GatewayClientError('server', envelope.error.code);
    } catch (error) {
      if (error instanceof GatewayClientError) throw error;
      throw new GatewayClientError('invalid_response');
    }
  }
}
